### Fase 1: Pré-processamento e Carga dos Dados

#### 1.1 Importação das bibliotecas essenciais

In [1]:
# Importando as bibliotecas
import pandas as pd
import plotly.express as px

print("Bibliotecas importadas com sucesso!")

Bibliotecas importadas com sucesso!


#### 1.2 Carregamento dos dados

In [2]:
# Definindo o caminho do arquivo csv
caminho_arquivo = '../data/raw/base_churn_preditiva_20191030.csv'

# Lendo o arquivo, especificando que o separador é ';'
df = pd.read_csv(caminho_arquivo, sep=';')

print("As 5 primeiras linhas do DataFrame:")
display(df.head())

As 5 primeiras linhas do DataFrame:


ClientId DataExtracao  Score_Credito          Estado    Gênero  Idade  \
0   345.568   2019-06-30            619       São Paulo  Feminino     42   
1   345.569   2019-06-30            608  Rio de Janeiro  Feminino     41   
2   345.570   2019-06-30            502       São Paulo  Feminino     42   
3   345.571   2019-06-30            699       São Paulo  Feminino     39   
4   345.572   2019-06-30            850  Rio de Janeiro  Feminino     43   

   Tempo_Cliente Limite_Credito_Mercado  Qte_Categorias  Usa_Cartao_Credito  \
0              2                      0               1                   1   
1              1               838,0786               1                   0   
2              8              1.596,608               3                   1   
3              1                      0               2                   0   
4              2             1.255,1082               1                   1   

   Programa_Fidelidade Sum_Pedidos_Acumulados DataUltimaTransacao  Churn  
0                    1                422,287          2019-09-14      1  
1                    1         468,9274166667          2019-10-05      0  
2                    0             474,714875          2019-08-20      1  
3                    0         390,9442916667          2019-10-21      0  
4                    1         329,5170833333          2019-10-26      0

#### 1.3 Diagnóstico inicial dos dados

In [3]:
# 1. Qual o tamanho do nosso dataset? (Linhas, Colunas)
print(f"O DataFrame possui {df.shape[0]} linhas e {df.shape[1]} colunas.")

O DataFrame possui 10000 linhas e 14 colunas.


In [4]:
# 2. Quais os tipos de dados e há valores faltando?
print("Tipos de Dados e Contagem de Nulos por Coluna:")
df.info()

Tipos de Dados e Contagem de Nulos por Coluna:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   ClientId                10000 non-null  float64
 1   DataExtracao            10000 non-null  object 
 2   Score_Credito           10000 non-null  int64  
 3   Estado                  10000 non-null  object 
 4   Gênero                  10000 non-null  object 
 5   Idade                   10000 non-null  int64  
 6   Tempo_Cliente           10000 non-null  int64  
 7   Limite_Credito_Mercado  10000 non-null  object 
 8   Qte_Categorias          10000 non-null  int64  
 9   Usa_Cartao_Credito      10000 non-null  int64  
 10  Programa_Fidelidade     10000 non-null  int64  
 11  Sum_Pedidos_Acumulados  10000 non-null  object 
 12  DataUltimaTransacao     10000 non-null  object 
 13  Churn                   10000 non-null  int64

In [5]:
# 3. Como as variáveis numéricas se distribuem?
print("Estatísticas Descritivas das Variáveis Numéricas:")
display(df.describe())

Estatísticas Descritivas das Variáveis Numéricas:


ClientId  Score_Credito         Idade  Tempo_Cliente  \
count  10000.000000   10000.000000  10000.000000   10000.000000   
mean     350.567500     650.528800     38.921800       5.012800   
std        2.886896      96.653299     10.487806       2.892174   
min      345.568000     350.000000     18.000000       0.000000   
25%      348.067750     584.000000     32.000000       3.000000   
50%      350.567500     652.000000     37.000000       5.000000   
75%      353.067250     718.000000     44.000000       7.000000   
max      355.567000     850.000000     92.000000      10.000000   

       Qte_Categorias  Usa_Cartao_Credito  Programa_Fidelidade         Churn  
count    10000.000000         10000.00000         10000.000000  10000.000000  
mean         1.530200             0.70550             0.515100      0.201300  
std          0.581654             0.45584             0.499797      0.400992  
min          1.000000             0.00000             0.000000      0.000000  
25%          1.000000             0.00000             0.000000      0.000000  
50%          1.000000             1.00000             1.000000      0.000000  
75%          2.000000             1.00000             1.000000      0.000000  
max          4.000000             1.00000             1.000000      1.000000

In [6]:
# 4. Como as variáveis categóricas se distribuem?
# Selecionando apenas as colunas que são do tipo 'object' (texto)
df_categorico = df.select_dtypes(include="object")

# Aplicando o describe nesse novo DataFrame filtrado
print("Estatísticas Descritivas das Variáveis Categóricas:")
display(df_categorico.describe())

Estatísticas Descritivas das Variáveis Categóricas:


DataExtracao     Estado     Gênero Limite_Credito_Mercado  \
count         10000      10000      10000                  10000   
unique            1          3          2                   6382   
top      2019-06-30  São Paulo  Masculino                      0   
freq          10000       5014       5457                   3617   

       Sum_Pedidos_Acumulados DataUltimaTransacao  
count                   10000               10000  
unique                   9999                 122  
top            103,8538333333          2019-10-12  
freq                        2                 296

Análise dos resultados do Diagnóstico

- As colunas DataExtracao, Limite_Credito_Mercado, Sum_Pedidos_Acumulados e DataUltimaTransacao estão com o Dtype (tipo de dado) como object, ou seja, estão sendo tratadas como texto. Isso nos impede de fazer qualquer cálculo matemático ou análise de data com elas.

- Sem Dados Faltantes: a coluna Non-Null Count mostra 10000 para todas as variáveis. Não será preciso preencher dados ausentes.

- Dimensões: temos 10.000 clientes e 14 variáveis para analisar.

- Idade: a idade dos clientes varia de 18 a 92 anos, com uma média de aproximadamente 39 anos.

- Estado: temos clientes em 3 estados, sendo São Paulo o mais frequente.

- Churn: A taxa de Churn é de 20.13% (mean de 0.2013 na coluna Churn), o que é um número bem significativo.

#### 1.4 Limpeza dos dados

In [7]:
# Lista das colunas numéricas que precisam de correção
cols_num_clean = ["Limite_Credito_Mercado", "Sum_Pedidos_Acumulados"]
for col in cols_num_clean:
    df[col] = df[col].astype(str).str.replace(".", "", regex=False)
    df[col] = df[col].str.replace(",", ".", regex=False).astype(float)

In [8]:
# Lista das colunas de data que precisam de correção
cols_date_clean = ["DataExtracao", "DataUltimaTransacao"]
for col in cols_date_clean:
    df[col] = pd.to_datetime(df[col])

In [9]:
print("\n--- Verificando os tipos de dados após a limpeza ---")
df.info()


--- Verificando os tipos de dados após a limpeza ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   ClientId                10000 non-null  float64       
 1   DataExtracao            10000 non-null  datetime64[ns]
 2   Score_Credito           10000 non-null  int64         
 3   Estado                  10000 non-null  object        
 4   Gênero                  10000 non-null  object        
 5   Idade                   10000 non-null  int64         
 6   Tempo_Cliente           10000 non-null  int64         
 7   Limite_Credito_Mercado  10000 non-null  float64       
 8   Qte_Categorias          10000 non-null  int64         
 9   Usa_Cartao_Credito      10000 non-null  int64         
 10  Programa_Fidelidade     10000 non-null  int64         
 11  Sum_Pedidos_Acumulados  10000 non-null  float64      

A saída do df.info() agora parece fazer sentido:

- datetime64[ns] para as datas.

- float64 para os valores monetários.

- object apenas para Estado and Gênero, que são de fato texto.

### Fase 2: Análise Exploratória de Dados (EDA)

#### 2.1 Análise Univariada

O objetivo aqui é entender o perfil do nosso cliente. Vamos responder a perguntas como:

- Qual é a distribuição de idade dos nossos clientes?

- Como o score de crédito está distribuído?

- Qual estado tem mais clientes?

In [18]:
# Mapeando valores binários (0/1) para rótulos descritivos

# Mapeiando a coluna 'Usa_Cartao_Credito'
map_cartao = {1: "Cadastrou", 0: "Não cadastrou"}
df["Usa_Cartao_Credito_Label"] = df["Usa_Cartao_Credito"].map(map_cartao)

# Mapeiando a coluna 'Programa_Fidelidade'
map_fidelidade = {1: "Participa", 0: "Não participa"}
df["Programa_Fidelidade_Label"] = df["Programa_Fidelidade"].map(map_fidelidade)

# Exibindo as primeiras linhas para verificar a transformação
display(
    df[
        [
            "Usa_Cartao_Credito",
            "Usa_Cartao_Credito_Label",
            "Programa_Fidelidade",
            "Programa_Fidelidade_Label",
        ]
    ].head()
)

Usa_Cartao_Credito Usa_Cartao_Credito_Label  Programa_Fidelidade  \
0                   1                Cadastrou                    1   
1                   0            Não cadastrou                    1   
2                   1                Cadastrou                    0   
3                   0            Não cadastrou                    0   
4                   1                Cadastrou                    1   

  Programa_Fidelidade_Label  
0                 Participa  
1                 Participa  
2             Não participa  
3             Não participa  
4                 Participa

In [19]:
# Lista das variáveis categóricas de interesse
variaveis_categoricas = [
    "Estado",
    "Gênero",
    "Usa_Cartao_Credito_Label",
    "Programa_Fidelidade_Label",
]

# Loop para criar a tabela de frequência e o gráfico de barras para cada variável
for var in variaveis_categoricas:
    print(f"\n---------- Análise da Variável: '{var}' ----------")

    # 1. Tabela de Frequência
    tabela_frequencia = df[var].value_counts().reset_index()
    tabela_frequencia.columns = [var, "Contagem"]
    tabela_frequencia["Porcentagem (%)"] = (
        tabela_frequencia["Contagem"] / len(df) * 100
    ).round(1)

    print("Tabela de Frequência:")
    display(tabela_frequencia)

    # 2. Gráfico de Barras
    fig = px.bar(
        tabela_frequencia,
        x=var,
        y="Contagem",
        text=tabela_frequencia["Porcentagem (%)"].apply(lambda x: f"{x}%"),
        title=f'Distribuição da Variável "{var}"',
    )

    fig.show()


---------- Análise da Variável: 'Estado' ----------
Tabela de Frequência:


Estado  Contagem  Porcentagem (%)
0       São Paulo      5014             50.1
1    Minas Gerais      2509             25.1
2  Rio de Janeiro      2477             24.8


---------- Análise da Variável: 'Gênero' ----------
Tabela de Frequência:


Gênero  Contagem  Porcentagem (%)
0  Masculino      5457             54.6
1   Feminino      4543             45.4


---------- Análise da Variável: 'Usa_Cartao_Credito_Label' ----------
Tabela de Frequência:


Usa_Cartao_Credito_Label  Contagem  Porcentagem (%)
0                Cadastrou      7055             70.6
1            Não cadastrou      2945             29.4


---------- Análise da Variável: 'Programa_Fidelidade_Label' ----------
Tabela de Frequência:


Programa_Fidelidade_Label  Contagem  Porcentagem (%)
0                 Participa      5151             51.5
1             Não participa      4849             48.5

In [11]:
# Lista de variáveis numéricas
variaveis_numericas = [
    "Score_Credito",
    "Idade",
    "Tempo_Cliente",
    "Limite_Credito_Mercado",
    "Qte_Categorias",
    "Sum_Pedidos_Acumulados",
]

# Loop para criar a tabela de estatísticas e o histograma para cada variável
for var in variaveis_numericas:
    print(f"\n---------- Análise da Variável: '{var}' ----------")

    # 1. Tabela de Estatísticas Descritivas
    print("Estatísticas Descritivas:")
    display(
        df[[var]].describe().transpose()
    ) 

    # 2. Histograma
    fig = px.histogram(
        df, x=var, title=f'Distribuição da Variável "{var}"', marginal="box"
    )

    fig.show()


---------- Análise da Variável: 'Score_Credito' ----------
Estatísticas Descritivas:


count      mean        std    min    25%    50%    75%    max
Score_Credito  10000.0  650.5288  96.653299  350.0  584.0  652.0  718.0  850.0


---------- Análise da Variável: 'Idade' ----------
Estatísticas Descritivas:


count     mean        std   min   25%   50%   75%   max
Idade  10000.0  38.9218  10.487806  18.0  32.0  37.0  44.0  92.0


---------- Análise da Variável: 'Tempo_Cliente' ----------
Estatísticas Descritivas:


count    mean       std  min  25%  50%  75%   max
Tempo_Cliente  10000.0  5.0128  2.892174  0.0  3.0  5.0  7.0  10.0


---------- Análise da Variável: 'Limite_Credito_Mercado' ----------
Estatísticas Descritivas:


count        mean         std  min  25%       50%  \
Limite_Credito_Mercado  10000.0  764.858893  623.974052  0.0  0.0  971.9854   

                              75%        max  
Limite_Credito_Mercado  1276.4424  2508.9809


---------- Análise da Variável: 'Qte_Categorias' ----------
Estatísticas Descritivas:


count    mean       std  min  25%  50%  75%  max
Qte_Categorias  10000.0  1.5302  0.581654  1.0  1.0  1.0  2.0  4.0


---------- Análise da Variável: 'Sum_Pedidos_Acumulados' ----------
Estatísticas Descritivas:


count        mean         std      min         25%  \
Sum_Pedidos_Acumulados  10000.0  417.042666  239.627053  0.04825  212.508792   

                               50%         75%      max  
Sum_Pedidos_Acumulados  417.474646  622.451031  833.302

#### 2.2 Principais Insights da Análise Univariada

Perfil Demográfico e Comportamental (Variáveis Categóricas)

* **Estado:** A base é fortemente concentrada em **São Paulo**, que representa **50.1%** de todos os clientes. O restante é dividido de forma quase idêntica entre **Minas Gerais (25.1%)** e **Rio de Janeiro (24.8%)**.
    
* **Gênero:** A distribuição entre gêneros é ligeiramente desequilibrada, com uma leve maioria do público **Masculino (54.6%)** em relação ao **Feminino (45.4%)**.
    
* **Uso de Cartão de Crédito:** A grande maioria dos clientes (**70.6%**) confirma que **"Sim (Cadastrou)"** um cartão de crédito no aplicativo. Apenas 29.4% indicam "Não (Não Cadastrou)".
    
* **Programa de Fidelidade:** A adesão ao programa de fidelidade é **extremamente equilibrada**, com **51.5%** dos clientes na categoria **"Sim (Participa)"** e **48.5%** na categoria **"Não (Não Participa)"**. Não há uma tendência dominante aqui.